core/src/sumeragi/message.rs

```diff
@@ -143,15 +139,49 @@ impl From<CommittedBlock> for BlockCommitted {
 #[non_exhaustive]
 pub struct BlockSyncUpdate {
     /// The corresponding block.
-    pub block: VersionedCandidateCommittedBlock,
+    block: VersionedCommittedBlock,
 }
 
 impl From<VersionedCommittedBlock> for BlockSyncUpdate {
     fn from(block: VersionedCommittedBlock) -> Self {
-        Self {
-            block: block.into(),
+        Self { block }
+    }
+}
+
+impl BlockSyncUpdate {
+    /// Extract block from block sync update message.
+    ///
+    /// # Errors
+    /// - When the block is invalid.
+    pub fn validate_and_extract_block(
+        self,
+        transaction_validator: &TransactionValidator,
+        wsv: &WorldStateView,
+        latest_block: Option<HashOf<VersionedCommittedBlock>>,
+        block_height: u64,
+    ) -> Result<VersionedCommittedBlock, eyre::Report> {
+        if block_height == 0 {
+            self.block.revalidate::<true>(
+                transaction_validator,
+                wsv,
+                latest_block,
+                block_height,
+            )?;
+            Ok(self.block)
+        } else {
+            self.block.revalidate::<false>(
+                transaction_validator,
+                wsv,
+                latest_block,
+                block_height,
+            )?;
+            Ok(self.block)
         }
     }
+    /// Bypass typesystem enforced validation.
+    pub fn extract_block_unchecked(self) -> VersionedCommittedBlock {
```

Comment:
This method is redundant, it's used in only in single place where we can just call `validate_and_extract_block` two times (we can chain them using [`or_else`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else)).

Reply:
The chaining causes borrow checker problems but I will make your change nonetheless.